src/chunkformatter.js

```diff
@@ -0,0 +1,319 @@
+/*!
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+const Mutation = require('./mutation.js');
+/**
+ * Enum for chunk formatter Row state.
+ * NEW_ROW: inital state or state after commitRow or resetRow
+ * ROW_IN_PROGRESS: state after first valid chunk without commitRow or resetRow
+ * CELL_IN_PROGRESS: state when valueSize > 0(partial cell)
+ */
+const RowStateEnum = Object.freeze({
+  NEW_ROW: 1,
+  ROW_IN_PROGRESS: 2,
+  CELL_IN_PROGRESS: 3,
+});
+/**
+ * ChunkFormatter formats all incoming chunks in to row
+ * keeps all intermediate state until end of stream.
+ * Should use new instance for each request.
+ */
+function ChunkFormatter() {
+  if (!(this instanceof ChunkFormatter)) {
+    return new ChunkFormatter();
+  }
+  this.reset();
+}
+/**
+ * Formats the row chunks into friendly format. Chunks contain 3 properties:
+ *
+ * `rowContents` The row contents, this essentially is all data pertaining
+ *     to a single family.
+ *
+ * `commitRow` This is a boolean telling us the all previous chunks for this
+ *     row are ok to consume.
+ *
+ * `resetRow` This is a boolean telling us that all the previous chunks are to
+ *     be discarded.
+ *
+ * @public
+ *
+ * @param {chunk[]} chunks The list of chunks.
+ * @param {object} [options] Formatting options.
+ * @param {callback} callback callback will be called with parsed rows
+ * @throws Error if it detects invalid state
+ * @example
+ * ChunkFormatter.formatChunks(chunks);
+ * // {
+ * //   follows: {
+ * //     gwashington: [
+ * //       {
+ * //         value: 2
+ * //       }
+ * //     ]
+ * //   }
+ * // }
+ */
+ChunkFormatter.prototype.formatChunks = function(chunks, options, callback) {
+  options = options || {};
+  let shouldContinue;
+  for (const chunk of chunks) {
+    switch (this.state) {
+      case RowStateEnum.NEW_ROW:
+        shouldContinue = this.newRow(chunk, options, callback);
```

Comment:
Is it possible to use an instance variable to track if it should continue? This is not an obvious pattern-- passing a callback and receiving a value that doesn't match the purpose of the method.